File: kmp/composeApp/src/commonMain/kotlin/com/tanyudii/tmuxweb/domain/BellAlert.kt
package com.tanyudii.tmuxweb.domain

/**
 * Direct port of public/notify.js's `shouldPlayBellAlert`/`buildBellTitle` —
 * see plan §2.6's "Bell alert" gap: today Web does title-flash + beep +
 * `Notification`, iOS does haptic-only. This pure function is the shared
 * decision logic both platforms drive their own delivery from. Kept
 * dependency-free like the JS original — `now`/`lastAlertAt` are passed in
 * rather than read from a clock, so it stays a pure function under test.
 */
const val BELL_COOLDOWN_MS = 1500L

fun buildBellTitle(sessionName: String?): String {
    val label = if (sessionName.isNullOrEmpty()) "session" else sessionName
    return "🔔 $label needs you — tmux-web"
}

fun shouldPlayBellAlert(
    muted: Boolean,
    hasFocus: Boolean,
    hidden: Boolean,
    lastAlertAt: Long?,
    now: Long,
): Boolean {
    if (muted) return false

    val isAway = hidden || !hasFocus
    if (!isAway) return false

    return lastAlertAt == null || now - lastAlertAt >= BELL_COOLDOWN_MS
}

File: web/src/domain/bellAlert.ts
// Ports kmp/composeApp/.../domain/BellAlert.kt (itself a direct port of the
// pre-KMP public/notify.js's shouldPlayBellAlert/buildBellTitle). Pure
// function: `now`/`lastAlertAt` are passed in rather than read from a
// clock, so it stays testable without faking Date.now().
export const BELL_COOLDOWN_MS = 1500;

export function buildBellTitle(sessionName: string | null): string {
  const label = sessionName ? sessionName : "session";
  return `🔔 ${label} needs you — tmux-web`;
}

export interface ShouldPlayBellAlertInput {
  muted: boolean;
  hasFocus: boolean;
  hidden: boolean;
  lastAlertAt: number | null;
  now: number;
}

export function shouldPlayBellAlert(input: ShouldPlayBellAlertInput): boolean {
  if (input.muted) return false;

  const isAway = input.hidden || !input.hasFocus;
  if (!isAway) return false;

  return input.lastAlertAt === null || input.now - input.lastAlertAt >= BELL_COOLDOWN_MS;
}
